### ISSUE TITLE
Requesting stylesheet over HTTP instead of HTTPS

### ISSUE THREAD

**Issue by wouterDeneyer** [firefly-iii/firefly-iii#921]
Hi, 

I have Firefly-III running behind an NGINX reverse proxy. 
I tried to upgrade to 4.6.7, but had to roll back. 

Firefly-iii is requesting stylesheets etc over a non-HTTPS connection. 
I have this in my .env file (and it is copied over from version to version)

```
APP_ENV=production
APP_DEBUG=false
APP_FORCE_SSL=true
APP_FORCE_ROOT=https://domainname.tk/firefly/
APP_KEY=base64:[key]
APP_LOG=daily
APP_LOG_LEVEL=warning
APP_URL=
```

What could be wrong? 

Thanks

**Comment by JC5** [comment]
Laravel 5.5, the underlying framework, has improved support for reverse proxies. Check out this previous issue that discusses how to fix it:

https://github.com/firefly-iii/firefly-iii/issues/882

You can use env variable `USE_PROXIES` to configure the $proxies variable.

**Comment by pkoziol** [comment]
Strange... I'm using nginx as reverse proxy to PHP (only these two, no other nginx-es nor Apache) and didn't have any problems on any patch version since 4.6.4.

My `.env` file:
```
APP_ENV=production
APP_DEBUG=true
APP_FORCE_SSL=true
APP_FORCE_ROOT=
APP_KEY=base64:[key]
APP_LOG=daily
APP_LOG_LEVEL=debug
APP_URL=https://firefly.example.org
```

**Comment by wouterDeneyer** [comment]
I'll try with JC5's links

Strange indeed. 
My reverse proxy is setup for 9 webservers(/services). 
All other 8 are working perfectly, so I 'ont think its due to my config. Also its working fine in 4.6.5.

`location /firefly {
            proxy_pass "https://firefly.freedom/firefly-iii/public/";
            proxy_set_header Host $host;
            proxy_set_header X-Real-IP $remote_addr;
            proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
         }
`

**Comment by nicoschreiner** [comment]
I think the problem is, that the latest version of Laravel doesn't read the proxy_headers by default and since Firefly-III 4.6.6 the Laravel-Version was updated (correct me, if I'm wrong).

Now we have to tell Laravel that it should read the proxy_headers. This can be done - as JC5 mentioned - by setting the env variable `USE_PROXIES` to `**` (at least this worked for my case :smiley: ). If you know the IP of your proxy you can set the `USE_PROXIES` to this IP and it should work.

**Comment by wouterDeneyer** [comment]
its `TRUSTED_PROXIES` indeed. 

However, I'm having bigger problems now. 
Now it's 404'ing, will have to investigate that first

**Comment by pkoziol** [comment]
```
proxy_pass "https://firefly.freedom/firefly-iii/public/";
```

So you have different setup than mine - you have some second HTTP server. My nginx serves non-php files staight from filesystem and php files via php-fpm. Thanks for that I don't need to setup `*_PROXIES` variables 😛 

```
server {
        server_name firefly.example.org;
        listen 443 ssl;

        include snippets/letsencrypt.conf;
        include snippets/ssl.conf;
        ssl_certificate            /etc/letsencrypt/live/firefly.example.org/fullchain.pem;
        ssl_certificate_key        /etc/letsencrypt/live/firefly.example.org/privkey.pem;

        access_log  /var/log/nginx/firefly.example.org/access.log;
        error_log   /var/log/nginx/firefly.example.org/error.log;

        root /home/firefly/firefly.example.org/public;
        index index.php;

        location / {
                try_files $uri $uri/ /index.php$is_args$args;
        }

        location ~* \.php$ {
                fastcgi_pass 127.0.0.1:9071;
                include      snippets/fastcgi-php.conf;
        }
}
```

**Comment by wouterDeneyer** [comment]
404's are "fixed". 
While testing, I must have removed a "/" in the `location /firefly/` line. 

For some reason, the URL gets re-written now. 

NGINX is configured to listen on port 80, 5443 and 443. 
80 redirects to HTTPS
5443 is a port-share between OpenVPN on pfSense and NGINX (used when accessing from WAN)
443 is only used on LAN and VPN access 
```
server {
    listen 80;
    return 301 https://$host$request_uri;
}

server {
    listen 443;
    listen 5443;
    server_name [servernames + LAN ip];
   [SSL configuration]

location /firefly/ {
            proxy_pass http://firefly.freedom/;
            proxy_set_header Host $host;
            proxy_set_header X-Real-IP $remote_addr;
            proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
         }
```

Apache conf file on the firefly server
```
<VirtualHost *:80>
        ServerAdmin webmaster@localhost
        DocumentRoot /var/www/firefly-iii/public/
        ErrorLog ${APACHE_LOG_DIR}/firefly/error.log
        CustomLog ${APACHE_LOG_DIR}//firefly/access.log combined
</VirtualHost>
```

When I go to https://example.com/firefly/
I get redirected to https://example.com/login


Any idea's?

**Comment by pkoziol** [comment]
After quick looking at the code and some googling I would say that you have two options that might work:

1\. Configure SSL on the Apache (problably you will need to tell nginx to trust self-signed certificate somehow) and use `proxy_pass https://firefly.freedom/;`. That way PHP would see `scheme == https` and would generate correct URLs. Probably... Drawback: another SSL - additional CPU overhead.

1b. Maybe there is some option (like `proxy_* Scheme https;` or something) that could trick PHP without additional SSL.

2\. According to StackOverflow ([1](https://stackoverflow.com/questions/29912997/laravel-routes-behind-reverse-proxy), [2](https://stackoverflow.com/questions/24426423/laravel-generate-secure-https-url-from-route)) adding `URL::forceScheme('https');` in the `boot` method of  `AppServiceProvider` could help. @nicoschreiner said in #882 that he tried that and it didn't work, but that was before `TRUSTED_PROXIES`. Maybe he still has that code in his files and that's why it's working for him.

**Comment by 4oo4** [comment]
@pkoziol Just tried option 2, and it throws a 500 error:

```
#0 /var/www/html/firefly/vendor/laravel/framework/src/Illuminate/View/FileViewFinder.php(79): Illuminate\\View\\FileViewFinder->findInPaths('errors.FireflyE...', Array)
#1 /var/www/html/firefly/vendor/laravel/framework/src/Illuminate/View/Factory.php(128): Illuminate\\View\\FileViewFinder->find('errors.FireflyE...')
#2 /var/www/html/firefly/vendor/laravel/framework/src/Illuminate/Routing/ResponseFactory.php(69): Illuminate\\View\\Factory->make('errors.FireflyE...', Array)
#3 /var/www/html/firefly/app/Exceptions/Handler.php(59): Illuminate\\Routing\\ResponseFactory->view('errors.FireflyE...', Array, 500)
#4 /var/www/html/firefly/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(326): FireflyIII\\Exceptions\\Handler->render(Object(Illuminate\\Http\\Request), Object(Symfony\\Component\\Debug\\Exception\\FatalThrowableError))
#5 /var/www/html/firefly/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(124): Illuminate\\Foundation\\Http\\Kernel->renderException(Object(Illuminate\\Http\\Request), Object(Symfony\\Component\\Debug\\Exception\\FatalThrowableError))
#6 /var/www/html/firefly/public/index.php(55): Illuminate\\Foundation\\Http\\Kernel->handle(Object(Illuminate\\Http\\Request))
#7 {main}
"} 
```

I'm trying to host mine on a subdirectory (`/firefly`), if that matters. Here's my `.env`, nginx config has identical proxy headers as @wouterDeneyer . In my case, nginx is doing a reverse proxy to Apache running on the same machine on port 8080.

Dumb question, but does `APP_URL` need to be the public URL like `APP_FORCE_ROOT`? I tried it with that as well as the backend URL (`http://localhost:8080/firefly/`) but got the same result.

```
APP_ENV=production
APP_DEBUG=false
APP_NAME=FireflyIII
APP_KEY=base64:xxxxxxxxxxxxxx
APP_LOG=daily
APP_LOG_LEVEL=warning
APP_URL=https://example.com/firefly/public/
TRUSTED_PROXIES=127.0.0.1
APP_FORCE_SSL=true
APP_FORCE_ROOT=https://example.com/firefly/public/
USE_PROXIES=127.0.0.1
```

**Comment by JC5** [comment]
As far as I know, the forceScheme('https') trick is superseded by the proxy middleware, which is why I removed it from Firefly III. The file exists (FireflyException.twig) but you could consider creating an empty file `FireflyException.blade.php`.

To make sure your logs capture every possible error (with more detail), change `APP_DEBUG` to true and set `APP_LOG_LEVEL` to `debug`

**Comment by wouterDeneyer** [comment]
I updated to the latest version today (4.6.8). 
Same problem. 

I enabled HTTPS with a self signed cert on the Firefly server, still the same :(

Weirdes thing: I have no logs in /storage/logs...

**Comment by pkoziol** [comment]
@wouterDeneyer you should probably add following code to your nginx configuration and there is a chance that HTTPS will be working.

```
proxy_set_header X-Forwarded-Proto $scheme;
```

@4oo4 looks like `APP_FORCE_*` were removed, so it doesn't matter what you put there.

**Comment by MacPaille** [comment]
Hey,

I'm a friend of Wouter and I did setup firefly on my raspberry pi with version 4.6.8. I have the exact same issue where the css files are downloaded using http instead of https.

I searched how to force laravel to use https but still couldn't find any good solution.
As my URL used in HTTPS is not accessible with regular HTTP on port 80 (my apache is listening on port 8001) it can't download the css even if I ask Chrome to load unsecured scripts.
